File: contest/LeetCode_Weekly_Contest_202/Magnetic_Force_Between_Two_Balls.java
class Solution {
    boolean check(int[] p, int dist, int m) {
        int n = p.length;
        int cnt = 1;
        int pre = p[0];
        for (int i = 1; cnt < m && i < n; ++i) {
            if (p[i] - pre >= dist) {
                cnt++;
                pre = p[i];
            }
        }
        return cnt >= m;
    }
    public int maxDistance(int[] p, int m) {
        Arrays.sort(p);
        int low = 0;
        int hight = 1000000000;
        int ans = 0;
        while (low <= hight) {
            int mid = (low + hight) >> 1;
            if (check(p, mid, m)) {
                ans = mid;
                low = mid + 1;
            } else {
                hight = mid - 1;
            }
        }
        return ans;
    }
}

File: contest/LeetCode_Weekly_Contest_202/Magnetic_Force_Between_Two_Balls.py
class Solution(object):
    def maxDistance(self, p, m):
        """
        :type position: List[int]
        :type m: int
        :rtype: int
        """
        p.sort()
        
        def check(dist):
            cnt, pre, n = 1, p[0], len(p)
            for i in xrange(1, n):
                if p[i] - pre >= dist:
                    pre = p[i]
                    cnt += 1
                if cnt == m:
                    return True
            return False
        
        low, hight, ans = 0, 1000000000, 0
        while low <= hight:
            mid = (low + hight) >> 1
            if check(mid):
                ans = mid
                low = mid + 1
            else:
                hight = mid - 1
        return ans
        
